xray/sql_context.go

```diff
@@ -0,0 +1,756 @@
+			return nil, ctx.Err()
+		}
+		dargs, err0 := namedValuesToValues(args)
+		if err0 != nil {
+			return nil, err0
+		}
+		err = Capture(ctx, stmt.attr.dbname, func(ctx context.Context) error {
+			stmt.populate(ctx)
+			var err error
+			result, err = stmt.Stmt.Query(dargs)
+			return err
+		})
+	}
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (stmt *driverStmt) ColumnConverter(idx int) driver.ValueConverter {
+	if conv, ok := stmt.Stmt.(driver.ColumnConverter); ok {
+		return conv.ColumnConverter(idx)
+	}
+	return driver.DefaultParameterConverter
+}
+
+func (stmt *driverStmt) populate(ctx context.Context) {
+	stmt.attr.populate(ctx, stmt.query)
+
+	seg := GetSegment(ctx)
+
+	if seg == nil {
+		processNilSegment(ctx)
+		return
+	}
+
+	seg.Lock()
+	seg.GetSQL().Preparation = "statement"
+	seg.Unlock()
+}
+
+// CheckNamedValue for implementing NamedValueChecker
+func (stmt *driverStmt) CheckNamedValue(nv *driver.NamedValue) (err error) {
+	if nvc, ok := stmt.Stmt.(namedValueChecker); ok {
+		return nvc.CheckNamedValue(nv)
+	}
+	// When converting data in sql/driver/convert.go, it is checked first whether the `stmt`
+	// implements `NamedValueChecker`, and then checks if `conn` implements NamedValueChecker.
+	// In the case of "go-sql-proxy", the `proxy.Stmt` "implements" `CheckNamedValue` here,
+	// so we also check both `stmt` and `conn` inside here.
+	if nvc, ok := stmt.conn.Conn.(namedValueChecker); ok {
+		return nvc.CheckNamedValue(nv)
+	}
+	// fallback to default
+	return defaultCheckNamedValue(nv)
+}
+
+func namedValuesToValues(args []driver.NamedValue) ([]driver.Value, error) {
+	var err error
+	ret := make([]driver.Value, len(args))
+	for _, arg := range args {
+		if len(arg.Name) > 0 {
+			err = errors.New("xray: driver does not support the use of Named Parameters")
+		}
+		ret[arg.Ordinal-1] = arg.Value
+	}
+	return ret, err
+}
+
+func stripPasswords(dsn string) string {
+	var (
+		tok        bytes.Buffer
+		res        bytes.Buffer
+		isPassword bool
+		inBraces   bool
+		delimiter  byte = ' '
+	)
+	flush := func() {
+		if inBraces {
+			return
+		}
+		if !isPassword {
+			res.Write(tok.Bytes())
+		}
+		tok.Reset()
+		isPassword = false
+	}
+	if strings.Count(dsn, ";") > strings.Count(dsn, " ") {
+		delimiter = ';'
+	}
+
+	buf := strings.NewReader(dsn)
+	for c, err := buf.ReadByte(); err == nil; c, err = buf.ReadByte() {
+		tok.WriteByte(c)
+		switch c {
+		case ':', delimiter:
+			flush()
+		case '=':
+			tokStr := strings.ToLower(tok.String())
+			isPassword = `password=` == tokStr || `pwd=` == tokStr
+			if b, err := buf.ReadByte(); err != nil {
+				break
+			} else {
+				inBraces = b == '{'
+			}
+			if err := buf.UnreadByte(); err != nil {
+				panic(err)
+			}
+		case '}':
+			b, err := buf.ReadByte()
+			if err != nil {
+				break
+			}
+			if b == '}' {
+				tok.WriteByte(b)
+			} else {
+				inBraces = false
+				if err := buf.UnreadByte(); err != nil {
+					panic(err)
+				}
+			}
+		case '@':
```

Comment:
Hi @shogo82148,

would you mind adding this updated logic in this PR? It will cover the case when it does not have the password and won't strip the user.


```
case '@':
			if strings.Contains(res.String(), ":") {
				resLen := res.Len()
				if resLen > 0 && res.Bytes()[resLen-1] == ':' {
					res.Truncate(resLen - 1)
				}
				isPassword = true
				flush()
				res.WriteByte(c)
			}
```